apisix/plugins/request-validation.lua

```diff
@@ -0,0 +1,84 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local core          = require("apisix.core")
+local plugin_name   = "request-validation"
+local json_decode   = require("cjson").decode
+local ngx           = ngx
+
+
+local schema = {
+    type = "object",
+    properties = {
+        body_schema = {type = "object"},
+        header_schema = {type = "object"}
+    },
+    oneOf = {
+        {required = {"body_schema"}},
+        {required = {"header_schema"}}
+    }
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 2800,
+    type = 'validation',
+    name = plugin_name,
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+function _M.rewrite(conf)
+    if conf.body_schema.properties.header_schema then
+        local headers = ngx.req.get_headers()
+        local ok, err = core.schema.check(conf.header_schema, headers)
+        if not ok then
+            core.log.error("req schema validation failed", err)
+            core.response.exit(400, err)
+        end
+    end
+
+    if not conf.body_schema.properties.body_schema then
+        ngx.req.read_body()
+        local body = ngx.req.get_body_data()
+        if not body then
+            body = ngx.req.get_body_file()
+        end
+
+        local data, error = json_decode(body)
```

Comment:
we should check the `Content-Type` header first.

https://stackoverflow.com/questions/20620300/http-content-type-header-and-json

Reply:
@membphis what would be the behavior if the header is not found, reject the call with `400`?

Reply:
https://stackoverflow.com/questions/9870523/differences-in-application-json-and-application-x-www-form-urlencoded

I think we can use default value: `application/json` if the header is not found.

if the header `Content-Type` is `application/x-www-form-urlencoded`, then we should use https://github.com/openresty/lua-nginx-module#ngxdecode_args .